File: client-ios/Sources/Network/AmbientClient.swift
import Foundation

/// HTTP client for the Ambient Intelligence network.
///
/// Implements the protocol flow:
/// 1. Get node public key (GET /pubkey)
/// 2. Encrypt prompt with node's key
/// 3. Submit job (POST /submit)
/// 4. Poll for completion (GET /status/{job_id})
/// 5. Decrypt response
///
/// Privacy: All prompts encrypted before leaving device.
@MainActor
class AmbientClient {
    private let nodeUrl: String
    private let crypto: AmbientCrypto
    private let session: URLSession
    
    init(nodeUrl: String) {
        self.nodeUrl = nodeUrl
        self.crypto = AmbientCrypto()
        
        // Configure session with no caching for privacy
        let config = URLSessionConfiguration.default
        config.requestCachePolicy = .reloadIgnoringLocalAndRemoteCacheData
        config.urlCache = nil
        self.session = URLSession(configuration: config)
    }
    
    /// Fetch node's public key for encryption.
    private func getNodePublicKey() async throws -> String {
        guard let url = URL(string: "\(nodeUrl)/pubkey") else {
            throw NetworkError.invalidURL
        }
        
        let (data, response) = try await session.data(from: url)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw NetworkError.httpError((response as? HTTPURLResponse)?.statusCode ?? 0)
        }
        
        let decoded = try JSONDecoder().decode(PublicKeyResponse.self, from: data)
        return decoded.publicKey
    }
    
    /// Submit encrypted job to node.
    private func submitJob(encryptedPrompt: String) async throws -> String {
        guard let url = URL(string: "\(nodeUrl)/submit") else {
            throw NetworkError.invalidURL
        }
        
        let requestBody = SubmitJobRequest(
            encryptedPrompt: encryptedPrompt,
            clientPubkey: try crypto.getPublicKeyB64()
        )
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(requestBody)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw NetworkError.httpError((response as? HTTPURLResponse)?.statusCode ?? 0)
        }
        
        let decoded = try JSONDecoder().decode(SubmitJobResponse.self, from: data)
        return decoded.jobId
    }
    
    /// Poll job status until complete.
    private func pollJobStatus(
        jobId: String,
        nodePubKey: String,
        maxWaitSeconds: Int = 300,
        pollIntervalSeconds: TimeInterval = 2.0
    ) async throws -> String {
        let startTime = Date()
        let maxWait = TimeInterval(maxWaitSeconds)
        
        while Date().timeIntervalSince(startTime) < maxWait {
            guard let url = URL(string: "\(nodeUrl)/status/\(jobId)") else {
                throw NetworkError.invalidURL
            }
            
            let (data, response) = try await session.data(from: url)
            
            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                throw NetworkError.httpError((response as? HTTPURLResponse)?.statusCode ?? 0)
            }
            
            let status = try JSONDecoder().decode(JobStatusResponse.self, from: data)
            
            switch status.status {
            case "complete":
                guard let encryptedResponse = status.encryptedResponse else {
                    throw NetworkError.noResponse
                }
                
                // Decrypt and return
                return try crypto.decryptFromNode(
                    encryptedB64: encryptedResponse,
                    nodePubKeyB64: nodePubKey
                )
                
            case "failed":
                throw NetworkError.jobFailed(status.errorMessage ?? "Unknown error")
                
            case "running":
                // Continue polling
                try await Task.sleep(nanoseconds: UInt64(pollIntervalSeconds * 1_000_000_000))
                
            default:
                throw NetworkError.unknownStatus(status.status)
            }
        }
        
        throw NetworkError.timeout
    }
    
    /// Submit prompt and wait for response.
    ///
    /// - Parameter prompt: User's plaintext prompt
    /// - Returns: Decrypted response from AI
    func submitAndWait(prompt: String) async throws -> String {
        // 1. Get node's public key
        let nodePubKey = try await getNodePublicKey()
        
        // 2. Encrypt prompt
        let encryptedPrompt = try crypto.encryptForNode(
            plaintext: prompt,
            nodePubKeyB64: nodePubKey
        )
        
        // 3. Submit job
        let jobId = try await submitJob(encryptedPrompt: encryptedPrompt)
        
        // 4. Poll until complete and decrypt
        return try await pollJobStatus(jobId: jobId, nodePubKey: nodePubKey)
    }
}

enum NetworkError: LocalizedError {
    case invalidURL
    case httpError(Int)
    case noResponse
    case jobFailed(String)
    case unknownStatus(String)
    case timeout
    
    var errorDescription: String? {
        switch self {
        case .invalidURL: return "Invalid URL"
        case .httpError(let code): return "HTTP error: \(code)"
        case .noResponse: return "Job complete but no response"
        case .jobFailed(let message): return "Job failed: \(message)"
        case .unknownStatus(let status): return "Unknown job status: \(status)"
        case .timeout: return "Job timed out"
        }
    }
}

File: client-android/src/main/java/com/ambient/client/network/AmbientClient.kt
package com.ambient.client.network

import com.ambient.client.crypto.AmbientCrypto
import kotlinx.coroutines.delay
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * HTTP client for the Ambient Intelligence network.
 *
 * Implements the protocol flow:
 * 1. Get node public key (GET /pubkey)
 * 2. Encrypt prompt with node's key
 * 3. Submit job (POST /submit)
 * 4. Poll for completion (GET /status/{job_id})
 * 5. Decrypt response
 *
 * Privacy: All prompts encrypted before leaving device.
 */
class AmbientClient(private val nodeUrl: String) : AutoCloseable {
    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        // Disable caching for privacy
        .cache(null)
        .build()

    private val crypto = AmbientCrypto()
    
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    /**
     * Fetch node's public key for encryption.
     */
    private suspend fun getNodePublicKey(): String {
        val request = Request.Builder()
            .url("$nodeUrl/pubkey")
            .get()
            .build()

        return httpClient.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw IOException("Failed to get node public key: HTTP ${response.code}")
            }
            
            val body = response.body?.string()
                ?: throw IOException("Empty response from node")
            
            val parsed = json.decodeFromString<PublicKeyResponse>(body)
            parsed.publicKey
        }
    }

    /**
     * Submit encrypted job to node.
     */
    private suspend fun submitJob(encryptedPrompt: String): String {
        val requestBody = SubmitJobRequest(
            encryptedPrompt = encryptedPrompt,
            clientPubkey = crypto.getPublicKeyB64()
        )
        
        val jsonBody = json.encodeToString(requestBody)
        val mediaType = "application/json; charset=utf-8".toMediaType()
        
        val request = Request.Builder()
            .url("$nodeUrl/submit")
            .post(jsonBody.toRequestBody(mediaType))
            .build()

        return httpClient.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw IOException("Failed to submit job: HTTP ${response.code}")
            }
            
            val body = response.body?.string()
                ?: throw IOException("Empty response from node")
            
            val parsed = json.decodeFromString<SubmitJobResponse>(body)
            parsed.jobId
        }
    }

    /**
     * Poll job status until complete.
     */
    private suspend fun pollJobStatus(
        jobId: String,
        nodePubKey: String,
        maxWaitSeconds: Int = 300,
        pollIntervalMs: Long = 2000
    ): String {
        val startTime = System.currentTimeMillis()
        val maxWaitMs = maxWaitSeconds * 1000L

        while (System.currentTimeMillis() - startTime < maxWaitMs) {
            val request = Request.Builder()
                .url("$nodeUrl/status/$jobId")
                .get()
                .build()

            httpClient.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    throw IOException("Failed to get job status: HTTP ${response.code}")
                }
                
                val body = response.body?.string()
                    ?: throw IOException("Empty response from node")
                
                val status = json.decodeFromString<JobStatusResponse>(body)
                
                when (status.status) {
                    "complete" -> {
                        val encryptedResponse = status.encryptedResponse
                            ?: throw IOException("Job complete but no response")
                        
                        // Decrypt and return
                        return crypto.decryptFromNode(encryptedResponse, nodePubKey)
                    }
                    
                    "failed" -> {
                        throw IOException("Job failed: ${status.errorMessage ?: "Unknown error"}")
                    }
                    
                    "running" -> {
                        // Continue polling
                        delay(pollIntervalMs)
                    }
                    
                    else -> {
                        throw IOException("Unknown job status: ${status.status}")
                    }
                }
            }
        }

        throw IOException("Job timed out after $maxWaitSeconds seconds")
    }

    /**
     * Submit prompt and wait for response.
     *
     * @param prompt User's plaintext prompt
     * @return Decrypted response from AI
     */
    suspend fun submitAndWait(prompt: String): String {
        // 1. Get node's public key
        val nodePubKey = getNodePublicKey()
        
        // 2. Encrypt prompt
        val encryptedPrompt = crypto.encryptForNode(prompt, nodePubKey)
        
        // 3. Submit job
        val jobId = submitJob(encryptedPrompt)
        
        // 4. Poll until complete and decrypt
        return pollJobStatus(jobId, nodePubKey)
    }

    override fun close() {
        crypto.close()
        // OkHttp client cleanup
        httpClient.dispatcher.executorService.shutdown()
        httpClient.connectionPool.evictAll()
    }
}
